gh-137197: Add SSLContext.set_ciphersuites to set TLS 1.3 ciphers

ronf opened the pull request:
This commit adds a new method SSLContext.set_ciphersuites which can be used to set TLS 1.3 cipher suites. It also updates the documentation, unit tests, and "what's new" text. A NEWS blurb will be added shortly.


<!-- gh-issue-number: gh-137197 -->
* Issue: gh-137197
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--137198.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

ronf committed:
gh-137197: Add SSLContext.set_ciphersuites to set TLS 1.3 ciphers

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

gpshead commented on Modules/_ssl.c:
```diff
@@ -3595,12 +3595,27 @@ _ssl__SSLContext_set_ciphers_impl(PySSLContext *self, const char *cipherlist)
 {
     int ret = SSL_CTX_set_cipher_list(self->ctx, cipherlist);
     if (ret == 0) {
-        /* Clearing the error queue is necessary on some OpenSSL versions,
```
We shouldn't remove this without understanding which OpenSSL versions it was referring to and whether they are still in use by CPython builds (1.1.x-ish API'd AWS-LC at a minimum, otherwise OpenSSL 3.0+).

ronf replied:
The clearing of the error queue is still happening here. I just took advantage of an existing helper function `setSSLError` to take care of this:

```python
static PyObject *
_setSSLError (_sslmodulestate *state, const char *errstr, int errcode, const char *filename, int lineno)
{
    if (errstr == NULL)
        errcode = ERR_peek_last_error();
    else
        errcode = 0;
    fill_and_set_sslerror(state, NULL, state->PySSLErrorObject, errcode, errstr, lineno, errcode);
    ERR_clear_error();
    return NULL;
}
```

picnixz replied:
Let's keep it that way and change it in a follow-up PR instead. Actually, the `errcode` parameter for `_setSSLError` is redundant as it's always overwritten.

gpshead commented on Doc/library/ssl.rst:
```diff
@@ -1684,19 +1684,29 @@ to speed up repeated connections from the same clients.
+   Set the allowed ciphers for sockets created with this context when
+   connecting using TLS 1.2 and earlier.  It should be a string in the `OpenSSL
+   cipher list format <https://docs.openssl.org/master/man1/ciphers/>`_.
+   To set allowed TLS 1.3 ciphers, use :meth:`SSHContext.set_ciphersuites`.
```
```suggestion
   To set allowed TLS 1.3 ciphers, use :meth:`SSLContext.set_ciphersuites`.
```

ronf replied:
Fixed - thanks for catching this.

ronf committed:
gh-137197: Fix typo in documentation

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -318,6 +318,11 @@ ssl
+  ciphers and updated the documentation on :meth:`ssl.SSLContext.set_ciphers`
+  to mention that it only applies to TLS 1.2 and earlier and that this new
+  method must be used to set TLS 1.3 cipher suites.
+  (Contributed by Ron Frederick in :gh:`137197`)
```
```suggestion
  (Contributed by Ron Frederick in :gh:`137197`)

```

ronf replied:
Fixed.

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -318,6 +318,11 @@ ssl
+* Added new method :meth:`ssl.SSLContext.set_ciphersuites` for setting TLS 1.3
+  ciphers and updated the documentation on :meth:`ssl.SSLContext.set_ciphers`
+  to mention that it only applies to TLS 1.2 and earlier and that this new
+  method must be used to set TLS 1.3 cipher suites.
```
```suggestion
* Added new method :meth:`ssl.SSLContext.set_ciphersuites` for setting TLS 1.3
  ciphers. For TLS 1.2 or earlier, use :meth:`ssl.SSLContext.set_ciphers` instead.
```

ronf replied:
What do you think about changing the text here to something like:

```
* Added new method :meth:`ssl.SSLContext.set_ciphersuites` for setting TLS 1.3
  ciphers. For TLS 1.2 or earlier, :meth:`ssl.SSLContext.set_ciphers` should
  continue to be used. Both calls can be made on the same context and the
  selected cipher suite will depend on the TLS version negotiated when a
  connection is made.
```

I wanted to somehow capture that these calls aren't mutually exclusive if you don't know in advance what version of TLS will end up being used. The existing call only affects cipher suites chosen when TLS 1.2 or earlier is negotiated, and the new call only affects cipher suites chosen when TLS 1.3 is negotiated.

ronf replied:
@picnixz, are you ok with this new wording?

ronf replied:
I've checked in the new wording I proposed. Let me know if you want any changes here.

I think this change may be ready to go if there are no other review comments.

picnixz replied:
Sorry, I forgot about this PR and I missed the notification I think. Let me have a look again.

ronf committed:
gh-137197: Update what's new text

Clarify when to use the original set_ciphers (TLS 1.2 and earlier)
vs. the new set_ciphersuites (TLS 1.3) methods and that both can
be used at once.

picnixz commented on Doc/library/ssl.rst:
```diff
@@ -1684,19 +1684,29 @@ to speed up repeated connections from the same clients.
-   <https://docs.openssl.org/master/man1/ciphers/>`_.
-   If no cipher can be selected (because compile-time options or other
+   Set the allowed ciphers for sockets created with this context when
+   connecting using TLS 1.2 and earlier.  It should be a string in the `OpenSSL
```
```suggestion
   connecting using TLS 1.2 and earlier.  *ciphers* should be a string in the `OpenSSL
```

I think it's better to say what's "it" is. I personally understand but removing any possible ambiguity would be better anyway. Up to you though. Same for *ciphersuites* below

ronf replied:
Done.

picnixz commented on Doc/library/ssl.rst:
```diff
@@ -1684,19 +1684,29 @@ to speed up repeated connections from the same clients.
+   connecting using TLS 1.2 and earlier.  It should be a string in the `OpenSSL
+   cipher list format <https://docs.openssl.org/master/man1/ciphers/>`_.
+   To set allowed TLS 1.3 ciphers, use :meth:`SSLContext.set_ciphersuites`.
+   below.  If no cipher can be selected (because compile-time options or other
```
```suggestion
   To set allowed TLS 1.3 ciphers, use :meth:`SSLContext.set_ciphersuites`.
   If no cipher can be selected (because compile-time options or other
```

ronf replied:
Fixed.

picnixz commented on Doc/library/ssl.rst:
```diff
@@ -1684,19 +1684,29 @@ to speed up repeated connections from the same clients.
    .. note::
       when connected, the :meth:`SSLSocket.cipher` method of SSL sockets will
-      give the currently selected cipher.
+      return the negotiated cipher and associated TLS version.
```
The docstring actually says:

```py
        """Return the currently selected cipher as a 3-tuple ``(name,
        ssl_version, secret_bits)``."""
```

So it's not a pair but a triplet that is being returned (that's how it's implemented in C). I suggest that we make this doc change in a separate PR.

ronf replied:
Do you want to revert just the changes made for the existing set_cipher() "note" but leave in this "note" text for the new set_ciphersuites() function? Alternately, what do you think of changing both function descriptions to say something like "return information about the negotiated cipher" and leave the the docs for the SSLSocket.cipher() method to cover this, where it already describes the tuple?

ronf replied:
After some thought, I'm thinking of the following for the notes text:

   .. note::
      when connected, the :meth:`SSLSocket.cipher` method of SSL sockets will
      return details about the negotiated cipher.

picnixz replied:
That's a good alternative as we don't need to know the details of `ciphers` in `set_ciphers/set_ciphersuites`. Another PR will take care of fixing the description of `ciphers`.

ronf replied:
The current docs for ``ciphers()`` says:

     Returns a three-value tuple containing the name of the cipher being used, the
     version of the SSL protocol that defines its use, and the number of secret
     bits being used.  If no connection has been established, returns ``None``.

This looks like it could be ok.

picnixz replied:
Oh.. so yes it's already fine!

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -261,10 +261,9 @@ def utc_offset(): #NOTE: ignore issues like #1647654
-                     cert_reqs=ssl.CERT_NONE, ca_certs=None,
-                     ciphers=None, certfile=None, keyfile=None,
-                     **kwargs):
+def test_wrap_socket(sock, *, cert_reqs=ssl.CERT_NONE, ca_certs=None,
```
Maybe revert the linebreak changes (namely keep `sock, *` on its own line) and just add `ciphersuites` and `min_version` after `ciphers`, and put `certfile` and `keyfile` on their own lines as well.

ronf replied:
WIll do.

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -2109,6 +2112,28 @@ def test_ciphers(self):
+                              min_version=ssl.TLSVersion.TLSv1_3) as s:
+            s.connect(self.server_addr)
+            self.assertEqual(s.cipher()[1], "TLSv1.3")
+        with test_wrap_socket(socket.socket(socket.AF_INET),
```
What about builds that don't support TLS 1.3?

ronf replied:
I'll add a `@requires_tls_version('TLSv1_3')` wrapper around this test.

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -2109,6 +2112,28 @@ def test_ciphers(self):
                                     cert_reqs=ssl.CERT_NONE, ciphers="^$:,;?*'dorothyx")
                 s.connect(self.server_addr)
 
+    def test_ciphersuites(self):
```
Shouldn't we ensure that this test is run for builds with TLS 1.3?

ronf replied:
Done.

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -2109,6 +2112,28 @@ def test_ciphers(self):
+            s.connect(self.server_addr)
+            self.assertEqual(s.cipher(),
+                             ("TLS_AES_256_GCM_SHA384", "TLSv1.3", 256))
+        # Error checking can happen at instantiation or when connecting
```
When does it happen upon instantiation and when does it happen while connecting? is it possible to only raise when we do `s.connect` or only raise when we create the socket?

ronf replied:
Passing in a bad cipher suite name would give an error when instantiating the SSL context (when set_ciphersuites() is called). So, the "XXX" case could be tested without a call to s.connect(). I can also add another test where the client & server have no overlapping TLS 1.3 cipher suites triggering a handshake failure during the s.connect().

picnixz replied:
Ok, for the XXX case, let's remove the call to connect(). And let's also test when we don't have overlapping cipher suites.

ronf replied:
Agreed. I have updated code for both of these cases - I should have it checked in shortly.

ronf committed:
gh-137197: Address review comments

picnixz commented on Doc/library/ssl.rst:
```diff
@@ -1684,19 +1684,30 @@ to speed up repeated connections from the same clients.
+
+   .. note::
+      when connected, the :meth:`SSLSocket.cipher` method of SSL sockets will
+      return details about the negotiated cipher.
```
```suggestion
   .. note::

      When connected, the :meth:`SSLSocket.cipher` method of SSL sockets will
      return details about the negotiated cipher.
```

You can also update the other note. I never saw that we had a missing capital letter.

ronf replied:
Fixed both instances.

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -332,6 +332,13 @@ ssl
+  continue to be used. Both calls can be made on the same context and the
+  selected cipher suite will depend on the TLS version negotiated when a
+  connection is made.
+  (Contributed by Ron Frederick in :gh:`137197`)
```
```suggestion
  (Contributed by Ron Frederick in :gh:`137197`.)
```

You don't need to update the entry just above. We'll clean it up in a PR for #133879.

picnixz replied:
This still needs to be addressed.

ronf replied:
Sorry - I misread this as saying the other PR would cover both this new text and the text above. I'll make the change just for the new entry for now.

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -263,7 +263,8 @@ def utc_offset(): #NOTE: ignore issues like #1647654
 def test_wrap_socket(sock, *,
                      cert_reqs=ssl.CERT_NONE, ca_certs=None,
-                     ciphers=None, certfile=None, keyfile=None,
+                     ciphers=None, ciphersuites=None, min_version=None,
```
Just for the future, but how about adding `max_version` as well? and put `min_version` and `max_version` on their own lines? even if we don't use it, I think it's fine to add it just in case.

ronf replied:
The current version of test_wrap_socket() only lets you specify min_version for now. It would be easy to add max, but it's called from 35 places. Would you change all of these to set both min & max?

The defaults for min/max version are special symbols MININUM_SUPPORTED and MAXIMUM_SUPPORTED, so there's no need to set the maximum unless you want to forcibly restrict it (in the future) to not using a TLS version greater than 1.3. I'm not sure I see a good reason to do that here, at least not yet. It would really come down to whether TLS 1.4 did something to yet again change the method used to set cipher suites, which I'm hoping won't be the case.

picnixz replied:
Actually, it was more to ease checks in internal code paths.

ronf replied:
Sorry, I'm not sure what you mean.

picnixz replied:
My suggestion was:

```py
def test_wrap_socket(sock, *,
                     cert_reqs=ssl.CERT_NONE, ca_certs=None,
                     ciphers=None, ciphersuites=None,
                     min_version=None, max_version=None,
                     ...):

    ...
    if min_version is not None:
        context.minimum_version = min_version
    if max_version is not None:
        context.maximum_version = max_version
    ...
```

That way, we can set `maximum_version` in tests if needed. For now, we don't but in the future we could. For instance, a test where we call `set_ciphersuites` but with `context.maximum_version = ssl.TLSVersion.TLSv1_2`. Or is the maximum version actually ignored?

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -1863,6 +1868,10 @@ class SimpleBackgroundTests(unittest.TestCase):
         self.server_context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
+
+        if has_tls_version('TLSv1_3'):
+            self.server_context.set_ciphersuites('TLS_AES_256_GCM_SHA384')
```
This could raise so we should be careful (for weird build options, I think it's possible to disable TLS_AES_256_GCM_SHA384). I would suggest only calling `set_ciphersuites` in tests that need it (that is those decorated by `requires_tls_version`. Otherwise the server_context would unconditoinally have this ciphersuite even if a test doesn't need it.

ronf replied:
I must admit I didn't love that change, but the problem here is in testing the mismatched ciphers case. I we leave the server accepting all ciphers, then there's no way to trigger a mismatched cipher, and I think the server is set up only once and then re-used for multiple test methods.

This mismatch test could be broken into its own test class so that we only set client and server cipher suites for that one specific test, but it would mean adding more boiler plate to set up the server in two different classes.

I felt pretty comfortable letting it be shared as this change doesn't impact the set of TLS 1.2 cipher suites used by any of these tests (either client or server) and any existing code would not be attempting to set any TLS 1.3 ciphers as that functionality didn't exist. Some of those tests might use TLS 1.3 today based on the defaults set but changing the server side to a specific cipher will still allow TLS 1.3 to be used, since the client-side would not be restricting TLS 1.3 ciphers.

That said, if you'd prefer I replicate the server setup to restrict this, I'm happy to do so.

As for the specific cipher I chose, I tried to pick the most "vanilla" one I could. For a test like this to exist, we're going to have to pick SOMETHING on the client and something else on the server to create the mismatch. I could try to do something like a get_ciphers(), filtering on TLS 1.3, and then pick two different values from that list, but it would complicate the test. For the unit testing, do you expect any of the test environments to disable these cipher suites? I see other test code hardcoding ciphers like these.

picnixz replied:
> and I think the server is set up only once and then re-used for multiple test methods.

The server context is new for every test though? setUp() is called before every test method. If you want to ease your life, just have different test methods, each of them will have their own `server_context` that is independent.

ronf replied:
I was thinking of setUpClass() instead of setUp(), which only runs once for all tests in a class. However, even with setUp() running before each test, I'd still need to somehow pass into setUp() whether to restrict the ciphers on a test-by-test basis, and I'm not sure how to do that.

Since there's not a lot of setup code, I'm going to see what it looks like if I add a new class, with its own separate setUp() function, used only for this mismatch test. That'll minimize changes made to the existing test code.

picnixz replied:
Ah yes, I see the issue. It's because we create the server before and changing its context afterwards wouldn't help right?

ronf replied:
Yes, that's right.

ronf replied:
Ok - I've checked in a new version of the test cases which uses its own class, reverting the changes to the previously modified class. This version also attempts to dynamically select ciphers from the list of available ciphers on the system. It assumes at least one cipher is available when the TLS 1.3 enabled check passes, but it will skip the mismatched cipher test if there's only one available TLS 1.3 cipher.

ronf committed:
gh-137197: Rework test cases

This commit reworks the set_ciphersuites() test cases, moving them into
their own class to avoid any changes to existing tests. It also makes
the cipher selection dynamic to avoid potentially trying to use a cipher
not available in some environments.

ronf committed:
gh-137197: Fix capitalization in docs

picnixz commented on Doc/library/ssl.rst:
```diff
@@ -1684,19 +1684,30 @@ to speed up repeated connections from the same clients.
 
-      TLS 1.3 cipher suites cannot be disabled with
-      :meth:`~SSLContext.set_ciphers`.
+.. method:: SSLContext.set_ciphersuites(ciphersuites)
```
This needs a `.. versionadded:: next` (sorry I forgot about it).

ronf replied:
Will do.

picnixz commented on Doc/library/ssl.rst:
```diff
@@ -1684,19 +1684,30 @@ to speed up repeated connections from the same clients.
+   connecting using TLS 1.2 and earlier.  The *ciphers* argument should
+   be a string in the `OpenSSL cipher list format
    <https://docs.openssl.org/master/man1/ciphers/>`_.
+   To set allowed TLS 1.3 ciphers, use :meth:`SSLContext.set_ciphersuites`.
```
Note to myself: I would have added a linebreak after this sentence to separate from "If no cipher..." but I think the entire page needs some edits, so let's not spend too much time on wording for now.

ronf replied:
Since this change is already changing the line right above that, I'll go ahead and add the blank line.

picnixz commented on Doc/library/ssl.rst:
```diff
@@ -2844,10 +2855,15 @@ TLS 1.3
+  called instead of :meth:`SSLContext.set_ciphers`, which only affects
+  ciphers in older TLS versions.  The method :meth:`SSLContext.get_ciphers`
+  returns information about ciphers for both TLS 1.3 and earlier versions
+  and the method :meth:`SSLSocket.cipher` returns information about the
```
In this paragraph, I would say "The XXX method" instead of "The method XXX".

ronf replied:
Will do.

picnixz commented on Doc/library/ssl.rst:
```diff
@@ -2844,10 +2855,15 @@ TLS 1.3
-  :meth:`SSLContext.set_ciphers` cannot enable or disable any TLS 1.3
-  ciphers yet, but :meth:`SSLContext.get_ciphers` returns them.
+- TLS 1.3 uses a disjunct set of cipher suites.  All AES-GCM and ChaCha20
+  cipher suites are enabled by default.  To restrict which TLS1.3 ciphers
```
```suggestion
  cipher suites are enabled by default.  To restrict which TLS 1.3 ciphers
```

ronf replied:
Thanks.

picnixz commented on Doc/library/ssl.rst:
```diff
@@ -2844,10 +2855,15 @@ TLS 1.3
-  ciphers yet, but :meth:`SSLContext.get_ciphers` returns them.
+- TLS 1.3 uses a disjunct set of cipher suites.  All AES-GCM and ChaCha20
+  cipher suites are enabled by default.  To restrict which TLS1.3 ciphers
+  are allowed, the method :meth:`SSLContext.set_ciphersuites` should be
```
For posterity: I really hate that OpenSSL named it `SSL_CTX_set_ciphersuites` for TLS 1.3 and later, and `SSL_CTX_set_cipher_list` for TLS 1.2 and below. I *hope* that this won't be annoying to users.

ronf replied:
I was surprised what they did here at first and I'm not a big fan about the naming chosen, but I can see why they didn't want to try and extend the existing command. It allows all kinds of partial matches on the five individual elements that make up a cipher suite in TLS 1.2 and earlier, plus other special keywords like LOW/MEDIUM/HIGH. All of that is gone with TLS 1.3, and the new function is much simpler and only supports an exact match against the very small number of cipher suites defined in TLS 1.3.

I don't really know why they didn't include "_list" in the name of the OpenSSL function, given that this seems to be the convention for other such calls that take a string argument.

The good news here is that things like post-quantum crypto are going to drive migration to TLS 1.3, and at some point setting TLS 1.2 cipher suites will no longer matter. At that point, the previous `set_ciphers()` can be deprecated and eventually removed, with only `set_ciphersuites()` remaining.

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -2238,6 +2243,53 @@ def test_transport_eof(self):
 
+@requires_tls_version('TLSv1_3')
+class SimpleBackgroundTestsTLS_1_3(unittest.TestCase):
+    """Tests that connect to a simple server running in the background"""
```
```suggestion
    """Tests that connect to a simple server running in the background."""
```

ronf replied:
Whether the doc-strings end in a period doesn't seem consistent, so I wasn't sure what the preference was. I'll go ahead and change this.

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -2238,6 +2243,53 @@ def test_transport_eof(self):
+                with self.assertRaises(ssl.SSLError):
+                    s.connect(self.server_addr)
+        else:
+            self.skipTest("Multiple TLS 1.3 ciphers are not available")
```
I think we should make a separate test method for this one, otherwise the entire test is marked as skipped (but the two first checks must unconditionally pass). You can also swap the `if` condition:

```py
def test_mismatched_ciphersuites(self):
    if self.matching_cipher == self.mismatched_cipher:
        self.skipTest("only one TLS 1.3 cipher is available")
    ...
```

ronf replied:
Will do.

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -2238,6 +2243,53 @@ def test_transport_eof(self):
+            self.assertEqual(s.cipher()[0], self.matching_cipher)
+
+        # Test mismatched TLS 1.3 cipher suites
+        if self.matching_client != self.mismatched_cipher:
```
`matching_client` should be `matching_cipher` right?

ronf replied:
Yes - thanks for catching this! I'm surprised I didn't get a test failure on that.

ronf committed:
gh-137197: Address review comments

picnixz commented on Doc/library/ssl.rst:
```diff
@@ -2844,10 +2858,15 @@ TLS 1.3
-  ciphers yet, but :meth:`SSLContext.get_ciphers` returns them.
+- TLS 1.3 uses a disjunct set of cipher suites.  All AES-GCM and ChaCha20
+  cipher suites are enabled by default.  To restrict which TLS 1.3 ciphers
+  are allowed, the method :meth:`SSLContext.set_ciphersuites` should be
```
```suggestion
  are allowed, the :meth:`SSLContext.set_ciphersuites` method should be
```

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -2238,6 +2246,53 @@ def test_transport_eof(self):
+                   if cipher['protocol'] == 'TLSv1.3']
+
+        self.matching_cipher = ciphers[0]
+        self.mismatched_cipher = ciphers[-1]
```
```suggestion
        if not ciphers:
            self.skipTest("no cipher supports TLSv1.3")

        self.matching_cipher = ciphers[0]
        # Some tests need at least two ciphers.
        self.mismatched_cipher = ciphers[-1]
```

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -2238,6 +2246,53 @@ def test_transport_eof(self):
+            with socket.socket(socket.AF_INET) as sock:
+                s = test_wrap_socket(sock, cert_reqs=ssl.CERT_NONE,
+                                     ciphersuites="XXX",
+                                     min_version=ssl.TLSVersion.TLSv1_3)
```
```suggestion
        with (
            socket.socket(socket.AF_INET) as sock,
            self.assertRaisesRegex(ssl.SSLError,
                                   "No cipher suite can be selected")
        ):
            s = test_wrap_socket(sock, cert_reqs=ssl.CERT_NONE,
                                 ciphersuites="XXX",
                                 min_version=ssl.TLSVersion.TLSv1_3)
```

That way, socket.socket() is not in the exception manager.

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -2238,6 +2246,53 @@ def test_transport_eof(self):
+                              ciphersuites=self.mismatched_cipher,
+                              min_version=ssl.TLSVersion.TLSv1_3) as s:
+            with self.assertRaises(ssl.SSLError):
+                s.connect(self.server_addr)
```
```suggestion
            self.assertRaises(ssl.SSLError, s.connect, self.server_addr)
```

picnixz commented on Misc/NEWS.d/next/Library/2025-07-29-05-12-50.gh-issue-137197.bMK3sO.rst:
```diff
@@ -0,0 +1 @@
+:mod:`ssl` can now set TLS 1.3 cipher suites.
```
```suggestion
:class:`~ssl.SSLContext` objects can now set TLS 1.3 cipher suites
via :meth:`~ssl.SSLContext.set_ciphersuites`.
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -3606,6 +3606,26 @@ _ssl__SSLContext_set_ciphers_impl(PySSLContext *self, const char *cipherlist)
+    if (ret == 0) {
+        _setSSLError(get_state_ctx(self), "No cipher suite can be selected.", 0, __FILE__, __LINE__);
+        return NULL;
+    }
```
```suggestion
    if (!SSL_CTX_set_ciphersuites(self->ctx, ciphersuites)) {
        _setSSLError(get_state_ctx(self), "No cipher suite can be selected.", 0, __FILE__, __LINE__);
        return NULL;
    }
```

picnixz reviewed (commented):
Just a few nitpicks.

ronf committed:
gh-137197: Address review comments

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -2238,6 +2246,67 @@ def test_transport_eof(self):
+class SimpleBackgroundTestsTLS_1_3(unittest.TestCase):
+    """Tests that connect to a simple server running in the background."""
+
+    @requires_tls_version('TLSv1_3')
```
I would suggest keeping it as a class decorator as the class name is about TLS 1.3

ronf replied:
I actually had trouble with the class decorator - that's why I moved it to the setUp() call. It looks like the `require_tls_version` decorator isn't written to to support being used as a class decorator. It doesn't report an error when you run the whole test_ssl_module, but it ends up not running the tests in the new class at all. If you try to run the tests on the specific class, it reports back an error:

```python
Traceback (most recent call last):
  File "/Users/ronf/src/cpython/Lib/runpy.py", line 198, in _run_module_as_main
    return _run_code(code, main_globals, None,
                     "__main__", mod_spec)
  File "/Users/ronf/src/cpython/Lib/runpy.py", line 88, in _run_code
    exec(code, run_globals)
    ~~~~^^^^^^^^^^^^^^^^^^^
  File "/Users/ronf/src/cpython/Lib/unittest/__main__.py", line 18, in <module>
    main(module=None)
    ~~~~^^^^^^^^^^^^^
  File "/Users/ronf/src/cpython/Lib/unittest/main.py", line 104, in __init__
    self.runTests()
    ~~~~~~~~~~~~~^^
  File "/Users/ronf/src/cpython/Lib/unittest/main.py", line 270, in runTests
    self.result = testRunner.run(self.test)
                  ~~~~~~~~~~~~~~^^^^^^^^^^^
  File "/Users/ronf/src/cpython/Lib/unittest/runner.py", line 257, in run
    test(result)
    ~~~~^^^^^^^^
  File "/Users/ronf/src/cpython/Lib/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
           ~~~~~~~~^^^^^^^^^^^^^^^
  File "/Users/ronf/src/cpython/Lib/unittest/suite.py", line 122, in run
    test(result)
    ~~~~^^^^^^^^
  File "/Users/ronf/src/cpython/Lib/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
           ~~~~~~~~^^^^^^^^^^^^^^^
  File "/Users/ronf/src/cpython/Lib/unittest/suite.py", line 122, in run
    test(result)
    ~~~~^^^^^^^^
  File "/Users/ronf/src/cpython/Lib/unittest/case.py", line 723, in __call__
    return self.run(*args, **kwds)
           ~~~~~~~~^^^^^^^^^^^^^^^
  File "/Users/ronf/src/cpython/Lib/unittest/case.py", line 644, in run
    testMethod = getattr(self, self._testMethodName)
AttributeError: 'SimpleBackgroundTestsTLS_1_3' object has no attribute 'runTest'. Did you mean: 'subTest'?
```

It'd probably be possible to modify `requires_tls_version` to support being a class decorator, but that seemed to be a much riskier change than just moving the decorator call to `setUp()`.

picnixz replied:
Oh! maybe add it to the class method `@classmethod def setUpClass(cls): pass` then

ronf replied:
What do you think of the following?

```patch
diff --git a/Lib/test/test_ssl.py b/Lib/test/test_ssl.py
index 7d79a50503..2f338509d1 100644
--- a/Lib/test/test_ssl.py
+++ b/Lib/test/test_ssl.py
@@ -2246,10 +2246,10 @@ def test_transport_eof(self):
             self.assertRaises(ssl.SSLEOFError, sslobj.read)


+@unittest.skipUnless(has_tls_version('TLSv1_3'), "TLS 1.3 is not available")
 class SimpleBackgroundTestsTLS_1_3(unittest.TestCase):
     """Tests that connect to a simple server running in the background."""

-    @requires_tls_version('TLSv1_3')
     def setUp(self):
         server_ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
         ciphers = [cipher['name'] for cipher in server_ctx.get_ciphers()
```

I found something similar to this wrapping the `TestPostHandshakeAuth` class.

picnixz replied:
This is also fine !

ronf replied:
Done!

picnixz reviewed (approved):
A final nitpick and we're good to go I think

ronf committed:
gh-137197: Address last review comment

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -332,6 +332,13 @@ ssl
 
    (Contributed by Ron Frederick in :gh:`136306`)
 
+* Added new method :meth:`ssl.SSLContext.set_ciphersuites` for setting TLS 1.3
```
```suggestion
* Added a new method :meth:`ssl.SSLContext.set_ciphersuites` for setting TLS 1.3
```

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -2238,6 +2246,67 @@ def test_transport_eof(self):
             self.assertRaises(ssl.SSLEOFError, sslobj.read)
 
 
+@unittest.skipUnless(has_tls_version('TLSv1_3'), "TLS 1.3 is not available")
```
Note to myself: change the `requires_tls_version` decorator so that it works.

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -2238,6 +2246,67 @@ def test_transport_eof(self):
+            self.skipTest("No cipher supports TLSv1.3")
+
+        self.matching_cipher = ciphers[0]
+        # Some tests need at least two ciphers.
```
```suggestion
        # Some tests need at least two ciphers, and are responsible
        # to skip themselves if matching_cipher == mismatched_cipher.
```

picnixz commented on Lib/test/test_ssl.py:
```diff
@@ -2238,6 +2246,67 @@ def test_transport_eof(self):
+                              min_version=ssl.TLSVersion.TLSv1_2,
+                              max_version=ssl.TLSVersion.TLSv1_2) as s:
+            s.connect(self.server_addr)
+            self.assertEqual(s.cipher()[1], 'TLSv1.2')
```
This test is asserting that the default `TLSv1_2` cipher suite is actually being used right? I'm fine with the current test but I'd like to fortify it (which I'll do) in the future where we have a ciphersuite for TLSv1.3 and a ciphersuite for TLSv1.2 and if we use min/max versions, then we're going to be using the TLSv1.2 ciphersuite we specified.

picnixz reviewed (approved):
I'm going to amend it directly and merge this.

picnixz committed:
Apply suggestions from code review
